mesonbuild/options.py

```diff
@@ -184,8 +200,10 @@ def from_string(cls, raw: str) -> 'OptionKey':
 
         return cls(opt, subproject, for_machine)
 
-    def evolve(self, name: T.Optional[str] = None, subproject: T.Optional[str] = None,
-               machine: T.Optional[MachineChoice] = None) -> 'OptionKey':
+    def evolve(self,
+               name: T.Optional[str] = _BAD_VALUE,
+               subproject: T.Optional[str] = _BAD_VALUE,
+               machine: T.Optional[MachineChoice] = _BAD_VALUE) -> 'OptionKey':
```

Comment:
There's a better pattern for this in the `KWargInfo.evolve()` method:

```python
class _NULL_T: pass
_NULL = _NULL_T()

def evolve(
        self, name: T.Union[str, _NULL_T] = _NULL,
        subproject: T.Union[str, None, _NULL_T] = _NULL,
        machine: T.Union[MachineChoice, _NULL_T] = _NULL) -> OptionKey:
    return type(self)(
        name if name is not _NULL else self.name,
        subproject is subproject is not _NULL else self.subproject
        machine if machine is not _NULL else self.machine)
```
The current annotations are invalid because the allow None in `name` and `machine`, but don't handle it.

This has the advantage of not trying to find an obscure enough string we don't hit it, it also means that we don't need to do string comparisons, but identity comparisons, it's either the `_NULL` instance, or it's not.

Reply:
This still needs to be addressed.